chainer/dataset/tabular/slice.py

```diff
@@ -0,0 +1,143 @@
+import numbers
+import numpy as np
+import six
+
+from chainer.dataset import TabularDataset
+
+
+class Slice(TabularDataset):
+
+    def __init__(self, dataset, indices, keys):
+        self._dataset = dataset
+        self._indices = _as_indices(indices, len(dataset))
+        self._key_indices = _as_key_indices(keys, dataset.keys)
+
+    def __len__(self):
+        if self._indices is None:
+            return len(self._dataset)
+        elif isinstance(self._indices, slice):
+            start, stop, step = self._indices.indices(len(self._dataset))
+            return len(six.moves.range(start, stop, step))
+        else:
+            return len(self._indices)
+
+    @property
+    def keys(self):
+        if self._key_indices is None:
+            return self._dataset.keys
+        else:
+            return tuple(self._dataset.keys[key_index]
+                         for key_index in self._key_indices)
+
+    @property
+    def mode(self):
+        return self._dataset.mode
+
+    def get_examples(self, indices, key_indices):
+        indices = _merge_indices(
+            self._indices, indices, len(self._dataset), len(self))
+        key_indices = _merge_key_indices(self._key_indices, key_indices)
+        return self._dataset.get_examples(indices, key_indices)
+
+
+class SliceHelper(object):
+
+    def __init__(self, dataset):
+        self._dataset = dataset
+
+    def __getitem__(self, args):
+        if isinstance(args, tuple):
+            indices, keys = args
+        else:
+            indices = args
+            keys = None
+
+        return Slice(self._dataset, indices, keys)
+
+
+def _as_indices(indices, len_):
```

Comment:
I profiled `_as_indices` by using MNIST example (same as #5954) and found that checking *all* values in `indices` degrade the performance. The following code improves by about 5%. I think we need a discussion about the validation spec around it.

```python
    if isinstance(indices, slice) or len(indices) == 0:
        return indices

    if isinstance(indices[0], (bool, np.bool_)):
        if not len(indices) == len_:
            raise ValueError('The number of booleans is '
                             'different from the length of dataset')
        return [i for i, index in enumerate(indices) if index]

    return [(i + len_ if i < 0 else i) for i in indices]
```

Reply:
I prefer strict checks since the mixture of ints and bools will be confusing for users.
Perhaps, we can provide an environment variable to disable strict checking for expert users (like `CHAINER_TYPE_CHECK`)

FYI: `numpy` checks only the first element of the list to determine the type of list. Actually, numpy accepts mixture of ints and bools.

Reply:
I think we can prevent the mixture by documentation. It seems Chainer checks just the head of list/array for its element in other cases (such like numpy). I like very much the strict manner too, but invoking `isinstance` for every item is little bit overkill in this situation.

Reply:
@beam2d How do you think of this?

Anyway, this modification should be in another PR.

Reply:
> FYI: numpy checks only the first element of the list to determine the type of list. Actually, numpy accepts mixture of ints and bools.

Sorry. This was wrong. numpy treat `indices` as index if at least one of elements is integer. Only if all elements are booleans, numpy treat it as booleans. So even if we make the behaviour numpy-compat, we need call `isinstance` for all elements.

Reply:
```python
np.arange(3)[[False, False, True]]  # array([2])
np.arange(3)[[False, False, 1]]  # array([0, 0, 1])
```

Reply:
@okapies I add mixed indices support in #7251. If the indices is a list of ints, `isinstance` is called only once.